File: android/src/main/kotlin/com/danjjohnson/compress_video/JobRegistry.kt
package com.danjjohnson.compress_video

import android.os.Handler
import java.io.File

/**
 * Main-thread-confined registry of live compression jobs, keyed by job id.
 *
 * Every method here must be called from the main thread -- the same thread every
 * `androidx.media3.transformer.Transformer` in this plugin is built and driven on
 * ([TransformerEngine], 02-RESEARCH.md Pattern 1). Main-thread confinement is the
 * synchronization here: there is no lock and no cross-thread mutable state.
 *
 * No Media3 or other Android-framework import beyond [Handler] -- [LiveJob] holds a
 * [LiveJob.cancelTransformer] callback rather than a `Transformer` reference directly, which
 * keeps this whole object exercisable with plain JVM test doubles (`CompressVideoPluginTest`'s
 * detach cases construct a [LiveJob] with no real `Transformer`, no Looper and no emulator --
 * `Transformer`'s own static initializer touches Android framework internals that are not
 * available outside a real device/emulator or Robolectric, so a real or mocked `Transformer`
 * instance cannot exist in that test at all).
 */
object JobRegistry {
    /**
     * One job's live state: how to cancel the operation driving it, the temp file it is writing
     * to, how to stop its progress polling, and how to resolve it when cancelled from outside
     * its own completion callback.
     */
    class LiveJob(
        val cancelTransformer: () -> Unit,
        val tempFile: File,
        val mainHandler: Handler,
        val progressRunnable: Runnable,
        val onCancelled: () -> Unit,
    ) {
        internal var cancelled: Boolean = false
    }

    private val jobs = LinkedHashMap<String, LiveJob>()

    /** Registers [job] under [jobId]. Must be called on the main thread. */
    fun register(
        jobId: String,
        job: LiveJob,
    ) {
        jobs[jobId] = job
    }

    /** Returns the live job for [jobId], or `null` if it is unknown or already finished. */
    fun find(jobId: String): LiveJob? = jobs[jobId]

    /**
     * Stops [jobId]'s progress polling and forgets it, without touching its files or invoking
     * its cancellation callback -- the normal path once a job's own completion callback has
     * already resolved it.
     */
    fun remove(jobId: String) {
        val job = jobs.remove(jobId) ?: return
        job.mainHandler.removeCallbacks(job.progressRunnable)
    }

    /**
     * Stops [jobId]'s progress polling WITHOUT forgetting it or touching its files -- called
     * from [TransformerEngine]'s `Transformer.Listener` terminal callbacks (`onCompleted`/
     * `onError`) the instant they fire, rather than relying on [remove] after the suspended
     * `compress` call resumes. The `Transformer` class javadoc states that `getProgress` reports
     * `PROGRESS_STATE_NOT_STARTED` once an export completes, so the polling loop must stop
     * itself proactively; stopping it here, synchronously inside the same main-Looper callback
     * that already fired, closes the (normally harmless, but unnecessary) window between the
     * listener firing and the coroutine's own cleanup running. A no-op if [jobId] is unknown.
     */
    fun stopPolling(jobId: String) {
        val job = jobs[jobId] ?: return
        job.mainHandler.removeCallbacks(job.progressRunnable)
    }

    /**
     * Cancels the job identified by [jobId]: stops its progress polling, cancels the underlying
     * export via [LiveJob.cancelTransformer], deletes its temp output file, forgets it, and
     * invokes its [LiveJob.onCancelled] callback so the suspended `startCompress` call can fail
     * with a typed cancelled error. A no-op if [jobId] is unknown or already cancelled -- so
     * cancelling twice, or cancelling after the job already finished (successfully or not) and
     * was removed from this registry, deletes nothing a second time and resolves without error.
     */
    fun cancel(jobId: String) {
        val job = jobs[jobId] ?: return
        if (job.cancelled) return
        job.cancelled = true
        job.mainHandler.removeCallbacks(job.progressRunnable)
        job.cancelTransformer()
        PluginFiles.quietDelete(job.tempFile)
        jobs.remove(jobId)
        job.onCancelled()
    }

    /** Cancels every live job -- used on plugin detach so no job outlives the engine (D-17). */
    fun cancelAll() {
        for (jobId in jobs.keys.toList()) {
            cancel(jobId)
        }
    }

    /**
     * Canonical paths of every live job's temp output file -- used by [PluginFiles.sweep] to
     * skip a file a still-running job is currently writing to, even when `clearCache()` runs
     * mid-job (T-02-28). Must be called on the main thread, exactly like every other method
     * here.
     */
    fun liveTempFilePaths(): Set<String> = jobs.values.map { it.tempFile.canonicalPath }.toSet()
}

File: darwin/compress_video/Sources/compress_video/JobRegistry.swift
import Foundation

/// Main-queue-confined registry of live compression jobs, keyed by job id.
///
/// Mirrors Android's `JobRegistry.kt`: every method here must be called from the main queue --
/// the same discipline Kotlin's version applies to the main Looper. Main-queue confinement is
/// the synchronization here: there is no lock and no cross-thread mutable state, matching the
/// Kotlin original exactly (it has none either).
///
/// `LiveJob` holds a `cancel` closure rather than a direct `AVAssetReader`/`AVAssetWriter`
/// reference -- the same reasoning as `JobRegistry.kt`'s own doc comment: this keeps the whole
/// object exercisable from plain XCTest with no simulator and no real AVFoundation objects in
/// scope at all (`RunnerTests.swift`'s `// MARK: - JobRegistry` section constructs a `LiveJob`
/// with a plain closure, no `AVAssetReader`/`AVAssetWriter` in sight).
enum JobRegistry {
  /// One job's live state: how to cancel the operation driving it, and the temp file it is
  /// writing to. A class (not a struct) so mutating `cancelled`/`terminal` through a dictionary
  /// lookup is visible to every holder of the same instance, mirroring Kotlin's `LiveJob` class.
  final class LiveJob {
    let cancel: () -> Void
    let tempFile: URL

    /// Set the instant `JobRegistry.cancel(jobId:)` actually cancels this job -- guards against
    /// invoking `cancel` a second time.
    fileprivate(set) var cancelled = false

    /// Set by `markTerminal(jobId:)` the instant `CompressionEngine`'s copy loop resolves this
    /// job's outcome (success or failure) -- BEFORE the suspended `compress()` call itself has
    /// resumed and called `remove(jobId:)` (mirrors `JobRegistry.kt`'s own WR-01 race-window
    /// note). A job that is `terminal` has already resolved its outcome on the native side;
    /// `cancel(jobId:)` must treat it exactly like an already-cancelled job and no-op, rather
    /// than deleting a temp file the still-suspended `compress()` call may be about to move
    /// into place.
    fileprivate(set) var terminal = false

    init(cancel: @escaping () -> Void, tempFile: URL) {
      self.cancel = cancel
      self.tempFile = tempFile
    }
  }

  private static var jobs: [String: LiveJob] = [:]

  /// Registers `job` under `jobId`. Must be called on the main queue.
  static func register(jobId: String, cancel: @escaping () -> Void, tempFile: URL) {
    jobs[jobId] = LiveJob(cancel: cancel, tempFile: tempFile)
  }

  /// Returns the live job for `jobId`, or `nil` if it is unknown or already finished.
  static func find(jobId: String) -> LiveJob? {
    jobs[jobId]
  }

  /// Forgets `jobId` -- the normal path once a job's own completion has already resolved it
  /// and moved (or discarded) its temp file. Does not touch any file itself.
  static func remove(jobId: String) {
    jobs.removeValue(forKey: jobId)
  }

  /// Marks `jobId` `LiveJob.terminal` WITHOUT forgetting it or touching its files -- called by
  /// `CompressionEngine`'s copy loop the instant a job's outcome (success or failure) resolves,
  /// closing the same race window `JobRegistry.kt`'s `stopPolling` closes (WR-01): a `cancel`
  /// call arriving after this point sees a job whose outcome is already decided and no-ops,
  /// rather than deleting a temp file the still-suspended `compress()` call may be about to move
  /// into place. A no-op if `jobId` is unknown.
  static func markTerminal(jobId: String) {
    jobs[jobId]?.terminal = true
  }

  /// Cancels the job identified by `jobId`: marks it cancelled, invokes its cancel closure
  /// (which `CompressionEngine` wires to flip its own local flag and call
  /// `reader.cancelReading()` -- observed by the copy loop on its next sample, per-sample
  /// latency being an accepted bound, D-08), and forgets it. A no-op if `jobId` is unknown,
  /// already cancelled, or already `terminal` -- so cancelling twice, or cancelling after the
  /// job's own outcome already resolved (successfully or not), invokes nothing and does not
  /// throw. Deletes no file itself -- `CompressionEngine`'s own copy loop, which alone holds
  /// the temp file's actual write handle, is responsible for that once it observes the flag
  /// and unwinds; touching the file from here (a different queue than the copy loop runs on)
  /// would race the writer still appending to it.
  static func cancel(jobId: String) {
    guard let job = jobs[jobId], !job.cancelled, !job.terminal else { return }
    job.cancelled = true
    job.cancel()
    jobs.removeValue(forKey: jobId)
  }

  /// Cancels every live job -- used by both platforms' teardown paths (iOS's
  /// `detachFromEngine(for:)`, macOS's `handleWillTerminate(_:)`) so no job outlives the plugin.
  static func cancelAll() {
    for jobId in Array(jobs.keys) {
      cancel(jobId: jobId)
    }
  }

  /// The temp file path of every currently-live job -- used by `PluginFiles.sweep` to skip a
  /// file a still-running job is currently writing to, even when `clearCache()` runs mid-job.
  /// Must be called on the main queue, exactly like every other method here.
  static func liveTempFilePaths() -> Set<String> {
    Set(jobs.values.map { $0.tempFile.resolvingSymlinksInPath().path })
  }
}
